kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -380,42 +392,55 @@ public void flushAndAwaitPendingEvents() {
       throw fatalException;
     }
 
-    // Make a (non-blocking) pass through the finalizing queue up front, to
-    // free up any recently-processed events before we attempt to drain the
-    // scheduling queue
-    drainFinalizingQueue();
-
-    while (!isCleared()) {
-      // we need to check this here because the check in finalizeAtLeastOneEvent only
-      // runs if there are currently slow events being handled by the async threads.
-      checkUncaughtExceptionsOnAsyncThreads();
```

Comment:
I essentially moved this check by putting it at the start of the `drainFinalizingQueue` method. It solves the problem mentioned here by ensuring that it always gets checked in finalizeAtLEastOneEvent.

It also helps with the check during punctuations and flushes, and to make sure that in case of an uncaught error (ie one that didn't make it into the finalizing queue), we abort quickly